x-pack/plugins/xpack_main/server/lib/xpack_usage.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import * as Rx from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+import { Cluster, ElasticsearchPlugin } from 'src/legacy/core_plugins/elasticsearch';
+import { Poller } from '../../../../common/poller';
+
+export type SecurityRealm =
+  | 'file'
+  | 'ldap'
+  | 'native'
+  | 'saml'
+  | 'kerberos'
+  | 'oidc'
+  | 'active_directory'
+  | 'pki';
+
+export interface XPackUsageResponse {
+  security: {
+    available: boolean;
+    enabled: boolean;
+    realms: {
+      [key in SecurityRealm]?: {
+        available: boolean;
+        enabled: boolean;
+      }
+    };
+    token_service: {
+      enabled: boolean;
+    };
+    api_key_service: {
+      enabled: boolean;
+    };
+    audit: {
+      enabled: boolean;
+    };
+  };
+}
+
+export interface XPackUsageContract {
+  getUsage$(): Rx.Observable<XPackUsageResponse | undefined>;
+  refreshNow(): void;
+}
+
+export interface XPackUsageDeps {
+  elasticsearch: ElasticsearchPlugin;
+}
+
+interface XPackUsageOptions {
+  pollFrequencyInMillis?: number;
+}
+
+type Logger = (...args: any[]) => void;
+
+export class XPackUsage {
+  private cluster: Cluster | null;
+
+  private poller: Poller | null;
+
+  private stop$ = new Rx.ReplaySubject(1);
+
+  private readonly pollFrequencyInMillis: number;
+
+  private readonly log: (tags: string[], message: string) => void;
+
+  private readonly usage$ = new Rx.BehaviorSubject<XPackUsageResponse | undefined>(undefined);
+
+  constructor(logger: Logger, { pollFrequencyInMillis = 30000 }: XPackUsageOptions) {
```

Comment:
It's not immediately clear to me the difference between what we should require in the `constructor`, vs. what should be considered a dependency that should be passed via `setup`, is this a convention in the new platform that I'm just not aware of?

Reply:
There is precedence for this in the new platform, but the distinction is hazy to me as well. @azasypkin do you know offhand what belongs in a service constructor vs service setup?

Reply:
TL;DR; `setup` - everything that makes sense and can be used only at the `setup` stage, `constructor` - everything that is valid during entire plugin lifetime.

---------------------

Plugin's `setup` method will receive "setup contracts" (basically what `setup` method of other plugins or core services returns) from all optional and required dependencies that live outside of the current plugin: either in other plugins (e.g. license or telemetry plugins) or in the core (e.g. elasticsearch or http services). Same applies to `stop` and `start` (that will eventually be re-introduced).

There is no notion of "plugin constructor" in the core, what core expects is a "initializer" function exported by the plugin that will receive "initializer" context. This context object will include everything that is valid and can be used during entire plugin lifetime (e.g. logger, env, config). Even though technically plugin doesn't have to define any classes, platform team recommends defining plugin as a class that is instantiated within "initializer" function and hence receives "initializer" context as a constructor argument + any other plugin specific stuff (assuming it's not expensive to create during "pre-setup" initialization).

Reply:
Everything above applies to plugins, but ideally services should follow similar pattern.

Reply:
That's a great explanation, thanks for clearing this up for me!